hollow/src/main/java/com/netflix/hollow/api/producer/AbstractHollowProducer.java

```diff
@@ -807,6 +807,13 @@ private void announce(ProducerListeners listeners, HollowProducer.ReadState read
         if (announcer != null) {
```

Comment:
Shouldn't this if clause be checked before we even fire the announcement listeners at line 808?

Reply:
If we moved the validation up, then between the validation happening and the announcement happening the leader lock could have been relinquished and a different leader might step in while this producer's announcement will be allowed to succeed based on a stale result.

But your comment helped me realize that the current implementation doesn't protect us entirely from the race condition, it just shortens the time window (now shortened to execution getting from line 810 to when the announcement is consistently available for reads) during which if the leader lock is released and a different instance acquires primary status then we'd have a bifurcated delta chain. More work is required.

Reply:
Added a revision that holds the leader lock for the time period between this check + until announcement is made. This should protect from the case where `producer.enablePrimaryProducer(false)` comes in after the primary check passes but before the announcement is made.

Reply:
Please take another look, and also inviting thoughts from @akhaku

Reply:
I think we additionally want to check before uploading uploading blobs since that could cause a delta chain overwrite

Reply:
I should probably elaborate so it's more clear: at least in our internal implementation of the delta chain, it's the act of uploading the delta and reverse delta blob that creates the delta chain (since we store to_version and from_version), not the announcement itself. The announcement only decides the "latest" version, the metadata on the blobs decides the delta chain, so if we wait until the announcement to check the "primary producer" status we're probably too late.

Reply:
The lock part looks good. I am just wondering why do we even care about firing the listeners if the announcement shouldn't be allowed to go through in case of a non-primary producer. That is why I was suggesting the check to be placed before we obtain the status of firing the announcement listeners.

Reply:
Sounds good @akhaku, I think a similar check at blob publish would be better as compared to extending the duration of the lock so that if someone's trying to disable primary they're not waiting forever (publish, listeners, etc.)

@nayanika-u this way the failure handling is a bit more clean. We call announcement start listeners, but we fail before announcing and we fire announcement complete with a failure status. It can be thought of as "the announcement failed because producer is not primary."

Failing before calling announce start would mean that we failed after validate but before announce. We'd need a way to represent to the user why we didn't get to announce stage i.e. we'd need a separate stage or maybe the check could be a part of the validation stage? But as I noticed earlier its good to reduce the time frame between when this check is done and when the announcement occurs (so for eg. a long running onAnnouncementStart listener wouldn't prolog that interval).

Reply:
Latest revision adds the check to delta and reverse delta publishing